internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,414 @@
+			return nil, err
+		}
+	}
+
+	jobHandle, err := windows.CreateJobObject(nil, jobName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			windows.Close(jobHandle)
+			delete(jobMap, uintptr(jobHandle))
+		}
+	}()
+
+	var mq *messageQueue
+	if notifications {
+		mq = newMessageQueue()
+		jobMap[uintptr(jobHandle)] = mq
+		if err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+			return nil, err
+		}
+	}
+
+	return &JobObject{
+		handle: jobHandle,
+		mq:     mq,
+	}, nil
+}
+
+// PollNotification will poll for a job object notification. This call should only be called once
+// per job (ideally in a goroutine loop) and will block if there is not a notification ready.
+// This call will return immediately with error `ErrNotRegistered` if the job was not registered
+// to receive notifications during `Create`. Internally, messages will be queued and there
+// is no worry of messages being dropped.
+func (job *JobObject) PollNotification() (interface{}, error) {
+	if job.mq == nil {
+		return nil, ErrNotRegistered
+	}
+	return job.mq.readOrWait()
+}
+
+// Close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *JobObject) Close() error {
+	job.handleLock.Lock()
+	defer job.handleLock.Unlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	if err := windows.Close(job.handle); err != nil {
+		return err
+	}
+
+	if job.mq != nil {
+		job.mq.close()
+	}
+	// Handles now invalid so if the map entry to receive notifications for this job still
+	// exists remove it so we can stop receiving notifications.
+	if _, ok := jobMap[uintptr(job.handle)]; ok {
+		delete(jobMap, uintptr(job.handle))
+	}
+
+	job.handle = 0
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *JobObject) Assign(pid uint32) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	if pid == 0 {
+		return errors.New("invalid pid: 0")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.handle, hProc)
+}
+
+// Terminate terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *JobObject) Terminate(exitCode uint32) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	return windows.TerminateJobObject(job.handle, exitCode)
+}
+
+// Shutdown gracefully shuts down all the processes in the job object. Context timeout
+// will be used as timer for graceful shutdown. If all of the processes in the job have not
+// exited by the timeout, the job will be terminated.
+func (job *JobObject) Shutdown(ctx context.Context) error {
+	job.handleLock.RLock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	job.handleLock.RUnlock()
```

Comment:
Is there really any reason to check this here? `Pids` and `Termintae` already check for this themselves.

Reply:
Yea thats very true. Too much copy paste from the other container types here